server/punishments.ts

```diff
@@ -108,87 +126,79 @@ class NestedPunishmentMap extends Map<RoomID, Map<string, Punishment>> {
 		}
 	}
 }
+
 /*********************************************************
- * Persistence
+ * Storage
  *********************************************************/
 
-export const Punishments = new class {
-	/**
-	 * ips is an ip:punishment Map
-	 */
-	ips = new PunishmentMap();
-	/**
-	 * userids is a userid:punishment Map
-	 */
-	userids = new PunishmentMap();
-	/**
-	 * roomUserids is a roomid:userid:punishment nested Map
-	 */
-	roomUserids = new NestedPunishmentMap();
-	/**
-	 * roomIps is a roomid:ip:punishment Map
-	 */
-	roomIps = new NestedPunishmentMap();
-	/**
-	 * sharedIps is an ip:note Map
-	 */
-	sharedIps = new Map<string, string>();
-	/**
-	 * Connection flood table. Separate table from IP bans.
-	 */
-	cfloods = new Set<string>();
-	/**
-	 * punishType is an allcaps string, for global punishments they can be
-	 * anything in the punishmentTypes map.
-	 *
-	 * This map can be extended with custom punishments by chat plugins.
-	 *
-	 * Keys in the map correspond to punishTypes, values signify the way
-	 * they should be displayed in /alt
-	 *
-	 */
-	punishmentTypes = new Map<string, string>([
-		['LOCK', 'locked'],
-		['BAN', 'globally banned'],
-		['NAMELOCK', 'namelocked'],
-	]);
-	/**
-	 * For room punishments, they can be anything in the roomPunishmentTypes map.
-	 *
-	 * This map can be extended with custom punishments by chat plugins.
-	 *
-	 * Keys in the map correspond to punishTypes, values signify the way they
-	 * should be displayed in /alt.
-	 * By default, this includes:
-	 * - 'ROOMBAN'
-	 * - 'BLACKLIST'
-	 * - 'BATTLEBAN'
-	 * - 'MUTE' (used by getRoomPunishments)
-	 *
-	 */
-	roomPunishmentTypes = new Map<string, string>([
-		['ROOMBAN', 'banned'],
-		['BLACKLIST', 'blacklisted'],
-		['BATTLEBAN', 'battlebanned'],
-		['MUTE', 'muted'],
-	]);
-	constructor() {
-		setImmediate(() => {
+class PunishmentsStorage {
+	static connect(type: string) {
+		let storage;
+		if (type === 'sqlite') {
+			storage = PunishmentsStorageSqlite;
+		} else if (type === 'tsv') {
+			storage = PunishmentsStorageTsv;
+		} else {
+			storage = PunishmentsStorageMemory;
+		}
+		storage.load();
+		return storage;
+	}
+}
+
+const PunishmentsStorageMemory = new class {
```

Comment:
This is a NoopPunishmentsStorage, not Memory. Though really, 'Storage' is a misnomer - nothing is being stored (yes, I understand we're already storing the information in JS elsewhere).

At bare minimum, the memory storage needs to store things (eg. you should be able to query in a test that the storage actually has the data stored).

Basically, right now the `Punishments` class *is* memory storage + additional logic on top. I'd separate out the storage bit, and thats the MemoryPunishmentsStorage implementation. You can then _compose_ the storage classes to get the same behavior regarding storing things in Memory and TSV/Sqlite